src/chainstate/stacks/index/trie_sql.rs

```diff
@@ -62,6 +66,8 @@ static SQL_MARF_DATA_TABLE: &str = "
 CREATE TABLE IF NOT EXISTS marf_data (
    block_id INTEGER PRIMARY KEY, 
    block_hash TEXT UNIQUE NOT NULL,
+   -- the trie itself.
+   -- if not used, then set to a zero-byte entry.
```

Comment:
It seems like we want a few of these fields to be NULLable, right?

If `data` isn't present, using a zero-byte entry is basically an approximation of "NULL", but with some extra weirdness involved. Similarly, `external_offset` and `external_length` using 0 rather than "NULL".

Reply:
I don't believe we can alter the schema to remove the `NOT NULL` constraint without copying the entire contents of the table to a new table with the constraint removed [1], so I'm not sure how valuable this particular change would be under the circumstances.

As for `external_offset` and `external_length`, sure -- these can be nullable.

[1] https://stackoverflow.com/questions/4007014/alter-column-in-sqlite

Reply:
On second thought, I'm going to push back on this.  In every case where we query `external_offset` and `external_length`, we're going to need to do `.unwrap_or(0)` anyway, as well as modify the `SELECT` queries to filter out rows where these values are `NULL`.  In order to do the latter efficiently, we're going to have to add an additional index for `external_length`, which would take more space than simply storing the default value `0` for these columns.

Reply:
That's fine. I don't think it's true that you'd need to filter out the NULL rows (the PR doesn't currently filter 0/0 rows, so why would you need to filter `NULL` rows?), and the `unwrap_or(0)` could be avoided by propagated this change upwards so that the structs _also_ represent lack of data as an `Option`, rather than a 0/0 (or like a storage type enum or something), but anyways, it isn't necessary, and if you want to push back on this refactor that works for me.

Reply:
> I don't think it's true that you'd need to filter out the NULL rows (the PR doesn't currently filter 0/0 rows, so why would you need to filter NULL rows?)

The PR depends on the ordering of `external_offset`, since the code needs to know the offset at which to append the next trie blob.  I don't know how stable the rules are in sqlite3 for comparing `NULL` to `INTEGER`, since on the face of it it ought to be undefined, but I don't want to find out later that `NULL` gets (re)defined as "greater than" the maximum `INTEGER` value in a `SELECT MAX(external_offset)` statement.